### ISSUE TITLE
Don't return "new" for empty catalog versions

### ISSUE THREAD

**Issue by jkosh44** [MaterializeInc/materialize#12092]
Currently in [storage.rs](https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/catalog/storage.rs) when `get_catalog_content_version` is called and there is no version, "new" is returned. I don't believe this is the correct value to return, and it just so happens by accident that it works out. Instead `None` should be returned and the caller should handle it.

Copying a comment from #11837 that explains it further.

> I see this method getting called in two places:
>
> 1. In `migrate`. 
> ```rust
> let catalog_version = storage.get_catalog_content_version()?;
>     let _catalog_version = match Version::parse(&catalog_version) {
>         Ok(v) => v,
>         // Catalog content versions changed to semver after 0.8.3, so all
>         // non-semver versions are less than that.
>         Err(_) => Version::new(0, 0, 0),
>     };
> ```
> Here it seems like `"new"` will get mapped to `Version::new(0,0,0)` which, at least according to the comment about version 0.8.3, doesn't seem right to me. The more I look at it actually the more I'm convinced that it will work out, but it seems like that's more by accident than on purpose. `"new"` will fail to parse and get mapped to `Version::new(0,0,0)` which will cause all migrations to be applied. Though it was very not obvious.
> 
> 2. In `open` for an error message
> ```rust
>             let last_seen_version = catalog.storage().get_catalog_content_version()?;
>             crate::catalog::migrate::migrate(&mut catalog).map_err(|e| {
>                 Error::new(ErrorKind::FailedMigration {
>                     last_seen_version,
>                     this_version: catalog.config().build_info.version,
>                     cause: e.to_string(),
>                 })
>             })?;
> ```
> This one makes more sense to me because the error message looks like this `#[error("cannot migrate from catalog version {last_seen_version} to version {this_version} (earlier versions might still work): {cause}")]`
> 
> It's a little hard to debug this method since it's only used for migrations and I deleted all the migrations (sorry about that). Though I > think it probably makes more sense for callers to handle a missing version number than for this method to map it to `"new"`.

**Comment by likzn** [comment]
@jkosh44 Hi,I want have a try this. Is something like below.
```rust
pub async fn get_catalog_content_version(&mut self) -> Result<String, Error> {
        self.get_setting("catalog_content_version")
            .await?
            .ok_or(Error::new(ErrorKind::GetSetting(
                "catalog_content_version".to_string(),
            )))
    }
```

**Comment by jkosh44** [comment]
Hi @likzn, thanks for your interest in this issue! I think you're on the right track, but the proper return type for this function should be `Result<Option<String, Error>>`. 

It's not an error if this value is missing, it just means that it hasn't been initialized yet. You can probably get rid of the call to `ok_or` here. The bigger change will be to update the callers of this function to handle `None` properly. Let me know if you need any guidance on that part.

**Comment by likzn** [comment]
@jkosh44 Thanks for explaining much for me! I think it is clear with your explanation. I will work it lately.

**Comment by jkosh44** [comment]
Fixed by #12646
